npm/translatePolicy.go

```diff
@@ -0,0 +1,1035 @@
+// Copyright 2018 Microsoft. All rights reserved.
+// MIT License
+package npm
+
+import (
+	"github.com/Azure/azure-container-networking/log"
+	"github.com/Azure/azure-container-networking/npm/iptm"
+	"github.com/Azure/azure-container-networking/npm/util"
+	networkingv1 "k8s.io/api/networking/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+type portsInfo struct {
+	protocol string
+	port     string
+}
+
+func craftPartialIptEntrySpecFromPort(portRule networkingv1.NetworkPolicyPort, sPortOrDPortFlag string) []string {
+	partialSpec := []string{}
+	if portRule.Protocol != nil {
+		partialSpec = append(
+			partialSpec,
+			util.IptablesProtFlag,
+			string(*portRule.Protocol),
+		)
+	}
+
+	if portRule.Port != nil {
+		partialSpec = append(
+			partialSpec,
+			sPortOrDPortFlag,
+			portRule.Port.String(),
+		)
+	}
+
+	return partialSpec
+}
+
+func craftPartialIptablesCommentFromPort(portRule networkingv1.NetworkPolicyPort, sPortOrDPortFlag string) string {
+	partialComment := ""
+	if portRule.Protocol != nil {
+		partialComment += string(*portRule.Protocol)
+		if portRule.Port != nil {
+			partialComment += "-"
+		}
+	}
+
+	if portRule.Port != nil {
+		partialComment += "PORT-"
+		partialComment += portRule.Port.String()		
+	}
+
+	if portRule.Protocol != nil || portRule.Port != nil {
+		partialComment += "-OF-"
+	}
+
+	return partialComment
+}
+
+func craftPartialIptEntrySpecFromOpAndLabel(op, label, srcOrDstFlag string, isNamespaceSelector bool) []string {
+	if isNamespaceSelector {
+		label = "ns-" + label
+	}
+	partialSpec := []string{
+		util.IptablesModuleFlag,
+		util.IptablesSetModuleFlag,
+		op,
+		util.IptablesMatchSetFlag,
+		util.GetHashedName(label),
+		srcOrDstFlag,
+	}
+
+	return util.DropEmptyFields(partialSpec)
+}
+
+func craftPartialIptEntrySpecFromOpsAndLabels(ns string, ops, labels []string, srcOrDstFlag string, isNamespaceSelector bool) []string {
+	var spec []string
+	
+	if len(ops) == 1 && len(labels) == 1 {
+		if ops[0] == "" && labels[0] == "" {
+			if !isNamespaceSelector {
+				// This is an empty podSelector,
+				// selecting all the pods within its namespace.
+				spec = []string{
+					util.IptablesModuleFlag,
+					util.IptablesSetModuleFlag,
+					util.IptablesMatchSetFlag,
+					util.GetHashedName("ns-" + ns),
+					srcOrDstFlag,
+				}
+			} else {
+				// This is an empty namespaceSelector,
+				// selecting all namespaces.
+				spec = []string{
+					util.IptablesModuleFlag,
+					util.IptablesSetModuleFlag,
+					util.IptablesMatchSetFlag,
+					util.GetHashedName(util.KubeAllNamespacesFlag),
+					srcOrDstFlag,
+				}
+			}
+
+			return spec
+		}
+	}
+
+	for i, _ := range ops {
+		spec = append(spec, craftPartialIptEntrySpecFromOpAndLabel(ops[i], labels[i], srcOrDstFlag, isNamespaceSelector)...)
```

Comment:
is ops and labels guaranteed to be same length?

Reply:
yes